public/app/features/transformers/regression/regression.ts

```diff
@@ -0,0 +1,112 @@
+import { PolynomialRegression } from 'ml-regression-polynomial';
+import { SimpleLinearRegression } from 'ml-regression-simple-linear';
+import { map } from 'rxjs';
+
+import {
+  DataFrame,
+  DataTransformerID,
+  FieldMatcherID,
+  FieldType,
+  SynchronousDataTransformerInfo,
+  fieldMatchers,
+} from '@grafana/data';
+
+export enum ModelType {
+  linear = 'linear',
+  polynomial = 'polynomial',
+}
+
+export interface RegressionTransformerOptions {
+  modelType?: ModelType;
+  degree?: number;
+  xFieldName?: string;
+  yFieldName?: string;
+  resolution?: number;
+}
+
+export const DEFAULTS = { resolution: 100, modelType: ModelType.linear, degree: 2 };
+
+export const RegressionTransformer: SynchronousDataTransformerInfo<RegressionTransformerOptions> = {
+  id: DataTransformerID.regression,
+  name: 'Regression',
+  operator: (options, ctx) => (source) =>
+    source.pipe(map((data) => RegressionTransformer.transformer(options, ctx)(data))),
+  transformer: (options, ctx) => {
+    return (frames: DataFrame[]) => {
+      const { resolution, modelType, degree } = { ...DEFAULTS, ...options };
+      if (frames.length === 0) {
+        return frames;
+      }
+      const matchesY = fieldMatchers.get(FieldMatcherID.byName).get(options.yFieldName);
+      const matchesX = fieldMatchers.get(FieldMatcherID.byName).get(options.xFieldName);
+
+      let xField;
+      let yField;
+      for (const frame of frames) {
+        const fy = frame.fields.find((f) => matchesY(f, frame, frames));
+        if (fy) {
+          yField = fy;
+          const fx = frame.fields.find((f) => matchesX(f, frame, frames));
+          if (fx) {
+            xField = fx;
+            break;
+          } else {
+            throw 'X and Y fields must be part of the same frame';
+          }
+        }
+      }
+
+      if (!xField || !yField) {
+        return frames;
+      }
+
+      // If x is a time field we normalize the time to the start of the timeseries data
+      const xFieldIsTime = xField.type === FieldType.time;
+
+      const lowest = Math.min(...xField.values);
+      const highest = Math.max(...xField.values);
```

Comment:
I think this can maybe be slightly optimised by just looping through the values once and calculating both min/max in one pass?

Reply:
I actually had to test this out. Interestingly enough the Math.min/max method is quite a bit faster (in chrome), unfortunately it blows the call stack if the array is too big :(

Will swap to a loop.

Reply:
Hmm, call stack blow is interesting. Good idea testing this out! 💯